### PR TITLE
Secure analytics endpoints with an API Key

### LINKED ISSUE #1977: Secure analytics endpoints with an API Key
Secure endpoints of `starter-analytics-postgres` endpoints with with [API Key](https://guides.micronaut.io/latest/micronaut-security-api-key.html) configuration.

### PR DESCRIPTION
**Pull request by guillermocalvo**
Closes #1977

### CODE REVIEW

File: starter-analytics-postgres/src/main/java/io/micronaut/starter/analytics/postgres/security/ApiKeyTokenValidator.java [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+    ApiKeyTokenValidator(ApiKeyRepository apiKeyRepository) {
+        this.apiKeyRepository = apiKeyRepository;
+    }
+
+    @Override
+    public Publisher<Authentication> validateToken(String token, HttpRequest<?> request) {
+        return Optional.ofNullable(request)
+                .map(HttpRequest::getPath)
+                .filter(path -> path.startsWith(AnalyticsController.PATH))
```
Comment (sdelamo) [inline starter-analytics-postgres/src/main/java/io/micronaut/starter/analytics/postgres/security/ApiKeyTokenValidator.java]:
```suggestion
                .filter(path -> path.startsWith(AnalyticsController.PATH)
```

it should apply to every path not just analytics

File: starter-analytics-postgres/src/test/groovy/io/micronaut/starter/analytics/postgres/StoreGeneratedProjectStatsSpec.groovy [hunk 2]
```diff
@@ -48,10 +52,50 @@ class StoreGeneratedProjectStatsSpec extends Specification implements TestProper
...
+                TestFramework.SPOCK,
+                MicronautJdkVersionConfiguration.DEFAULT_OPTION
+        )
+        generated.setSelectedFeatures([new SelectedFeature("google-cloud-function")])
+
+        when:
+        unauthorizedClient.applicationGenerated(generated).get()
+
+        then:
+        Exception e = thrown()
```
Comment (sdelamo) [inline starter-analytics-postgres/src/test/groovy/io/micronaut/starter/analytics/postgres/StoreGeneratedProjectStatsSpec.groovy]: use `HttpClientResponseException` which I think is the exception being thrown and check .getStatus() == 401

File: starter-analytics-postgres/src/test/groovy/io/micronaut/starter/analytics/postgres/StoreGeneratedProjectStatsSpec.groovy [hunk 3]
```diff
@@ -48,10 +52,50 @@ class StoreGeneratedProjectStatsSpec extends Specification implements TestProper
...
+                TestFramework.SPOCK,
+                MicronautJdkVersionConfiguration.DEFAULT_OPTION
+        )
+        generated.setSelectedFeatures([new SelectedFeature("google-cloud-function")])
+
+        when:
+        wrongApiKeyClient.applicationGenerated(generated).get()
+
+        then:
+        Exception e = thrown()
```
Comment (sdelamo) [inline starter-analytics-postgres/src/test/groovy/io/micronaut/starter/analytics/postgres/StoreGeneratedProjectStatsSpec.groovy]: use `HttpClientResponseException` which I think is the exception being thrown and check .getStatus() == 401

**Review by sdelamo — CHANGES_REQUESTED** [review]
starter-api consumes analytics api. We need to modify the code to supply an api via configuration and send it in the requests towards analytics.

File: starter-analytics-postgres/src/test/groovy/io/micronaut/starter/analytics/postgres/StoreGeneratedProjectStatsSpec.groovy [hunk 4]
```diff
@@ -73,8 +75,8 @@ class StoreGeneratedProjectStatsSpec extends Specification implements TestProper
         unauthorizedClient.applicationGenerated(generated).get()
 
         then:
-        Exception e = thrown()
-        e.cause.message == "Unauthorized"
+        ExecutionException e = thrown()
```
Comment (sdelamo) [inline starter-analytics-postgres/src/test/groovy/io/micronaut/starter/analytics/postgres/StoreGeneratedProjectStatsSpec.groovy:77]: Can you not use?

```suggestion
        HttpClientResponseException e = thrown()
```

File: starter-analytics-postgres/src/test/groovy/io/micronaut/starter/analytics/postgres/StoreGeneratedProjectStatsSpec.groovy [hunk 5]
```diff
@@ -92,8 +94,8 @@ class StoreGeneratedProjectStatsSpec extends Specification implements TestProper
         wrongApiKeyClient.applicationGenerated(generated).get()
 
         then:
-        Exception e = thrown()
-        e.cause.message == "Unauthorized"
+        ExecutionException e = thrown()
```
Comment (sdelamo) [inline starter-analytics-postgres/src/test/groovy/io/micronaut/starter/analytics/postgres/StoreGeneratedProjectStatsSpec.groovy:96]: Can you not use?

```suggestion
        HttpClientResponseException e = thrown()
```

File: starter-web-netty/src/test/groovy/io/micronaut/starter/netty/ReportAnalyticsSpec.groovy [hunk 6]
```diff
@@ -29,7 +30,7 @@ import java.util.concurrent.CompletableFuture
         value =  "http://localhost:8080/analytics/report")
 @Property(
         name = "micronaut.http.services.analytics.token",
-        value =  "false")
+        value =  "test-api-key")
```
Comment (sdelamo) [inline starter-web-netty/src/test/groovy/io/micronaut/starter/netty/ReportAnalyticsSpec.groovy:35]: can you add a non happy path test?

**Review by sdelamo — DISMISSED** [review]

File: starter-analytics-postgres/src/test/groovy/io/micronaut/starter/analytics/postgres/StoreGeneratedProjectStatsSpec.groovy (hunk 4 above)
Comment (guillermocalvo) [reply starter-analytics-postgres/src/test/groovy/io/micronaut/starter/analytics/postgres/StoreGeneratedProjectStatsSpec.groovy:77]: Not really. The HTTP client returns a `CompletableFuture` and therefore `HttpClientResponseException` is wrapped inside a `ExecutionException`.

File: starter-analytics-postgres/src/test/groovy/io/micronaut/starter/analytics/postgres/StoreGeneratedProjectStatsSpec.groovy (hunk 5 above)
Comment (guillermocalvo) [reply starter-analytics-postgres/src/test/groovy/io/micronaut/starter/analytics/postgres/StoreGeneratedProjectStatsSpec.groovy:96]: [Same here](https://github.com/micronaut-projects/micronaut-starter/pull/1980/files/c43ea85f87b37a1ef9377e450cce1f1e5182d2a9..a33d3c7099091b53317e74de4ff5e6b25d233f72#r1275001093).

File: starter-web-netty/src/test/groovy/io/micronaut/starter/netty/ReportAnalyticsSpec.groovy (hunk 6 above)
Comment (guillermocalvo) [reply starter-web-netty/src/test/groovy/io/micronaut/starter/netty/ReportAnalyticsSpec.groovy:35]: There are two non-happy-path tests already:
- [`test save generation data without api key`](https://github.com/micronaut-projects/micronaut-starter/pull/1980/files#diff-a127201424deff00e0a159547527155a6b80525c70ffdf3d02651d947330ca51R63)
- [`test save generation data with a wrong api key`](https://github.com/micronaut-projects/micronaut-starter/pull/1980/files#diff-a127201424deff00e0a159547527155a6b80525c70ffdf3d02651d947330ca51R82)

What other scenarios do you want to test, specifically?

Comment (sdelamo) [reply starter-web-netty/src/test/groovy/io/micronaut/starter/netty/ReportAnalyticsSpec.groovy:35]: I would like a a test which verifies how starter-web-netty behaves if the anlaytics server returns a 401?

Comment (sdelamo) [reply starter-web-netty/src/test/groovy/io/micronaut/starter/netty/ReportAnalyticsSpec.groovy:35]: the test you linked test analytics. I want tests for starter-web-netty

Comment (guillermocalvo) [reply starter-web-netty/src/test/groovy/io/micronaut/starter/netty/ReportAnalyticsSpec.groovy:35]: Fair enough. On it 👍

File: starter-web-netty/src/test/groovy/io/micronaut/starter/netty/ReportAnalyticsNoApiKeySpec.groovy [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+import io.micronaut.starter.options.BuildTool
+import io.micronaut.test.extensions.spock.annotation.MicronautTest
+import jakarta.inject.Inject
+import spock.lang.Specification
+
+@MicronautTest
+@Property(name = "spec.name", value = "ReportAnalyticsNoApiKeySpec")
+@Property(
+        name = "micronaut.http.services.analytics.url",
+        value = "http://localhost:8080/analytics/report")
```
Comment (sdelamo) [inline starter-web-netty/src/test/groovy/io/micronaut/starter/netty/ReportAnalyticsNoApiKeySpec.groovy]: this should not be `"http://localhost:8080/analytics/report` but `http://localhost:8080`

Comment (sdelamo) [reply starter-web-netty/src/test/groovy/io/micronaut/starter/netty/ReportAnalyticsNoApiKeySpec.groovy]: @guillermocalvo ping!

Comment (sdelamo) [inline starter-web-netty/src/test/groovy/io/micronaut/starter/netty/ReportAnalyticsNoApiKeySpec.groovy]:
```suggestion
        value = "http://localhost:8080")
```

**Review by sdelamo — DISMISSED** [review]

**Pull request merged**